### PR TITLE
CI: adopt matrix syntax

### PR DESCRIPTION
**Pull request by vemv**
## Brief

The project's build matrix predates CircleCI's matrix syntax. By using it, the build file can grow more sustainably - in particular we can add more elements / dimensions without having to review/adapt everything.

## QA plan

Build this branch and observe that each named job is different and actually results in different setups/commands being executed.

image

## Author checklist

* [x] I have QAed the functionality
* [x] The PR has a reasonably reviewable size and a meaningful commit history
* [ ] I have run the [branch formatter](https://github.com/nedap/formatting-stack/blob/332a419034ab46fad526a5592f4257353bd695b6/src/formatting_stack/branch_formatter.clj) and observed no new/significative warnings
* [x] The build passes
* [x] I have self-reviewed the PR prior to assignment
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [x] Correctness
  * [x] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [x] Documentation
  * [x] Security
  * [x] Performance
  * [x] Breaking API changes
  * [x] Cross-compatibility (Clojure/ClojureScript)

## Reviewer checklist

* [ ] I have checked out this branch and reviewed it locally, running it
* [ ] I have QAed the functionality
* [ ] I have reviewed the PR
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [ ] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes
  * [ ] Cross-compatibility (Clojure/ClojureScript)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -93,64 +93,33 @@ test_code_filters: &test_code_filters
       only: /^v\d+\.\d+\.\d+(-alpha\d+)?$/
 
 workflows:
-  version: 2.1
```
Comment (vemv) [inline .circleci/config.yml:96]: Removed because it's already at the top level

File: .circleci/config.yml [hunk 2]
```diff
@@ -93,64 +93,33 @@ test_code_filters: &test_code_filters
...
-          lein_test_command: lein with-profile -dev,+ci,+refactor-nrepl-latest do clean, test
-          <<: *test_code_filters
-      - test_code:
-          name: "JDK 11 excluding refactor-nrepl"
-          jdk_version: openjdk11
-          lein_test_command: lein with-profile -dev,+ci do clean, test
-          <<: *test_code_filters
+          matrix:
+            <<: *test_code_filters
+            alias: "test_code_parallel_eastwood"
```
Comment (vemv) [inline .circleci/config.yml:112]: By using two matrices, we can avoid useless combinations (parallel executor + non-parallel `lein test`), which in turn saves resources.

`alias` becomes needed when using two matrices backed by the same job

File: .circleci/config.yml [hunk 3]
```diff
@@ -93,64 +93,33 @@ test_code_filters: &test_code_filters
...
+            parameters:
+              jdk_version: ["openjdk8-parallel", "openjdk11-parallel"]
+              lein_test_command:
+                - "lein with-profile -dev,+ci,+refactor-nrepl,+parallel-eastwood do clean, test"
+                - "lein with-profile -dev,+ci,+refactor-nrepl-latest,+parallel-eastwood do clean, test"
```
Comment (vemv) [inline .circleci/config.yml:117]: Increased the matrix a little here (exercising `refactor-nrepl-latest`) also

**Comment by thumbnail** [comment]
Is there something we can do about the jobnames? The jobnames overflow in the CircleCI interface, so they required to be opened before it's clear what the job configuration is. It's kinda messy. 


Do like the transition to matrices though 🚀 !

**Comment by vemv** [comment]
> Is there something we can do about the jobnames? 

Following the docs, it looked like this would work:

```diff
diff --git a/.circleci/config.yml b/.circleci/config.yml
index be723e4..4982fe0 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -99,6 +99,7 @@ workflows:
           matrix:
             <<: *test_code_filters
             alias: "test_code_base"
+            name: "Serial - << matrix.jdk_version >> - << matrix.lein_test_command >>" 
             parameters:
               jdk_version: ["openjdk8","openjdk11"]
               lein_test_command:
@@ -109,6 +110,7 @@ workflows:
       - test_code:
           matrix:
             <<: *test_code_filters
+            name: "Parallel - << matrix.jdk_version >> - << matrix.lein_test_command >>"
             alias: "test_code_parallel_eastwood"
             parameters:
               jdk_version: ["openjdk8-parallel", "openjdk11-parallel"]
```

however I don't see a change in the UI

**Pull request closed without merging**